src/stapi_fastapi/products_router.py

```diff
@@ -0,0 +1,46 @@
+# Generic product router factory
+from typing import Self
+from fastapi import APIRouter, HTTPException, status, Request
+from stapi_fastapi.models.opportunity import OpportunityRequest
+from stapi_fastapi.models.product import Product
+from stapi_fastapi.backend import StapiBackend
+from stapi_fastapi.exceptions import ConstraintsException
+
+"""
+/products[MainRouter]/opportunities
+/products[MainRouter]/parameters
+/products[MainRouter]/order
+"""
+
+class ProductRouter(APIRouter):
+
+    def __init__(
+        self: Self,
+        product: Product,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(*args, **kwargs)
+        self.product = product
+
+        # TODO: use add_api_route
+        self.post("/opportunities", summary="Search Opportunities for the product")(self.search_opportunities)
```

Comment:
I _think_ we can type the endpoint to the product opportunity response model by using the `response_model` kwarg here. It seems like the `Product` class might have a method that returns a properly typed `OpportunitiesResponse` model using the constraints defined on the product.